lib/AST/DiagnosticEngine.cpp

```diff
@@ -1068,16 +1068,25 @@ void DiagnosticEngine::emitDiagnostic(const Diagnostic &diagnostic) {
 llvm::StringRef
 DiagnosticEngine::diagnosticStringFor(const DiagID id,
                                       bool printDiagnosticName) {
-  // TODO: Print diagnostic names from `localization`.
-  if (printDiagnosticName) {
-    return debugDiagnosticStrings[(unsigned)id];
-  }
   auto defaultMessage = diagnosticStrings[(unsigned)id];
+  auto diagnosticName = diagnosticNameStrings[(unsigned)id];
   if (localization) {
     auto localizedMessage =
         localization.get()->getMessageOr(id, defaultMessage);
+    if (printDiagnosticName) {
```

Comment:
I think it's a job of a localization producer to check that flag and produce expected message instead of diagnostic engine adding these postfixes explicitly. We should probably keep `debugDiagnosticStrings` around as a fallback too so we could do something like:

```c++
llvm::StringRef defaultMessage = printDiagnosticName 
                    ? debugDiagnosticStrings[(unsigned)id] 
                    : diagnosticNameStrings[(unsigned)id];
```

and pass it to `getMessageOr`.

Reply:
That would mean `getMessageOr` will take the `printDiagnosticName` boolean value and based on the `printDiagnosticName`'s value... if it's `true` will return the localized diagnostic message alongside the diagnostic name or just return the localized diagnostic message otherwise. Now, the localization producer is taking care of constructing the message whether it's for `debug` or not. Does that make sense?

Reply:
Can't localization producer just infer it in `setLocalization` instead?

Reply:
I'm sorry, I don't fully understand. Do you mean that we should construct the localization producers i.e. YAML and Serialized with `printDiagnosticNames`? If so, I'm wondering... why that approach would be better?

Reply:
I mean we could infer whether names should be printed or not during producer construction, that way we don't need to add any parameters to `getMessageOf`.

Reply:
Yes, by **infer**ing you mean passing the `printDiagnosticNames` to localization producers' constructors?

Reply:
Yes, that's right